src/storage/sectionStorage.ts

```diff
@@ -0,0 +1,121 @@
+
+const readSectionData = async (): Promise<ISectionData[]> => {
+  // versions <= v1.4 used the sync storagearea
+  const oldSections = await Browser.storage.sync.get("sections")
+  if (oldSections.sections !== undefined) {
+    console.log("Importing sections from sync storagearea...")
+    await Browser.storage.sync.remove("sections")
+    await Browser.storage.local.set({ sections: oldSections })
+  }
+
+  const rawSections = (await Browser.storage.local.get("sections")) as
+    | ValidVersionData
+    | VersionWithNoNumber
+  return processRawSections(rawSections)
+}
+
+const processRawSections = async (
+  rawSections: ValidVersionData | VersionWithNoNumber
+): Promise<ISectionData[]> => {
+  const extractedSections = isVersionWithNumber(rawSections)
+    ? rawSections
+    : manuallyDetermineVersion(rawSections)
+  return await sectionDataAutoMigrator(extractedSections)
+}
+
+/**
+ * Type-safe section data migration function. Recursively updates
+ * the data to newer and newer versions until it reaches current
+ * version.
+ */
+const sectionDataAutoMigrator = async (
+  input: ValidVersionData
+): Promise<ISectionData[]> => {
+  switch (input.version) {
+    case "2.0.1":
+      return input.data as ISectionData[]
+    case "1.6.0":
+    case "2.0.0":
+      return sectionDataAutoMigrator({
+        version: "2.0.1",
+        data: v2_0_0(input.data),
+      })
+    case "1.5.0":
+      return sectionDataAutoMigrator({
+        version: "2.0.0",
+        data: await v1_5_0(input.data),
+      })
+    case "1.4.1":
+      return sectionDataAutoMigrator({
+        version: "2.0.0",
+        data: await v1_4_1(input.data),
+      })
+    default:
+      throw Error(
+        // @ts-expect-error    this case shouldn't ever be possible, but on
+        // the off chance that it actually is, i'd like to get a runtime
+        // error for it
+        `Version ${input.version} could not be properly migrated! Are you sure this version of the extension supports this format?`
+      )
+  }
+}
+
+/**
+ * Convenience function to deserialize sections from JSON, using a
+ * custom reviver to convert JSONified `Set`s represented as arrays
+ * back to actual `Set`s.
+ */
+const loadSectionDataFromJSON = (
+  input: string
+): VersionWithNoNumber | ValidVersionData => {
+  return JSON.parse(input, (key: string, value: unknown) => {
+    if (key === "terms") {
+      return new Set((value as []).slice(1))
+    }
+    return value
+  })
+}
+
+/**
+ * Convenience function to serialize sections to JSON, using a custom
+ * replacer since `Set`s aren't JSON-able on their own.
+ */
+const convertSectionDataToJSON = (input: ISectionData[]): string => {
+  return JSON.stringify(
+    input,
+    (key: unknown, value: unknown) => {
+      if (value instanceof Set) {
+        return ["_isSet", ...value]
+      }
+      return value
+    },
+    2
+  )
+}
+
+const writeSectionData = async (newSections: ISectionData[]) => {
```

Comment:
Do we want to have this function here or in App.tsx since it's interacting with the webpage itself

Reply:
i'd like to keep it out of App.tsx since i feel like there's already a lot going on in there - this and `readSectionData` should stay together. they can probably go in their own little file, if you want to separate them from the migrator functions.

i'm not very opinionated on this one though so if you'd like to move it there let me know.

Reply:
Let's move it since in the future we might want to pull this folder out to be shared